### PR TITLE
executor: remove childrenResult from baseExecutor

### PR DESCRIPTION
**Pull request by lysu**
## What have you changed? (mandatory)

This is 1 of 3 PRs of "reduce chunk capacity". issue ref #7077

In this PR, we remove `childrenResult` from `baseExecutor` and push down them to real implements 

only create result chunk for children only when it's really needed, and try create single chunk if no needed to use more than one.

## What is the type of the changes? (mandatory)

- Improvement (non-breaking change which is an improvement to an existing feature), only code refactor.

## How has this PR been tested? (mandatory)

- unit tests
- integration tests

---
This change is [Reviewable](https://reviewable.io/reviews/pingcap/tidb/7076)

### CODE REVIEW

File: executor/aggregate.go [hunk 1]
```diff
@@ -795,6 +801,15 @@ func (e *StreamAggExec) Open(ctx context.Context) error {
 	return nil
 }
 
+// Close implements the Executor Close interface.
+func (e *StreamAggExec) Close() error {
+	if err := e.baseExecutor.Close(); err != nil {
```
Comment (zz-jason) [inline executor/aggregate.go]: We could firstly release the resources handled by itself, then release the resources handled by its children.

File: executor/projection.go [hunk 2]
```diff
@@ -216,6 +225,9 @@ func (e *ProjectionExec) prepare(ctx context.Context) {
 
 // Close implements the Executor Close interface.
 func (e *ProjectionExec) Close() error {
+	if e.isUnparallelExec() {
```
Comment (XuHuaiyu) [inline executor/projection.go]: why we need this check?

**Comment by shenli** [comment]
Is there a related issue about this PR?

File: executor/projection.go (hunk 2 above)
Comment (lysu) [reply executor/projection.go]: it paired for [Open](https://github.com/pingcap/tidb/pull/7076/files/56afa90414ebdb4caa97ceff1529993ceaf52e7b#diff-da03bd2739194a5e817f1b2a9685509bR77). It seems that alloc/dealloc child result chunk is required only when `isUnparallel==true`

**Comment by zz-jason** [comment]
LGTM

File: executor/aggregate.go [hunk 3]
```diff
@@ -795,6 +801,15 @@ func (e *StreamAggExec) Open(ctx context.Context) error {
 	return nil
 }
 
+// Close implements the Executor Close interface.
+func (e *StreamAggExec) Close() error {
+	e.childResult = nil
+	if err := e.baseExecutor.Close(); err != nil {
```
Comment (XuHuaiyu) [inline executor/aggregate.go]:
``` go
return errors.Trace(e.baseExecutor.Close())   // I prefer this.
```
or

``` go
err := e.baseExecutor.Close()
return errors.Trace(err)
```

File: executor/executor.go [hunk 4]
```diff
@@ -567,11 +563,21 @@ func (e *LimitExec) Open(ctx context.Context) error {
...
+	e.childResult = e.children[0].newChunk()
 	e.cursor = 0
 	e.meetFirstBatch = e.begin == 0
 	return nil
 }
 
+// Close implements the Executor Close interface.
+func (e *LimitExec) Close() error {
+	e.childResult = nil
+	if err := errors.Trace(e.baseExecutor.Close()); err != nil {
```
Comment (winoros) [inline executor/executor.go]: You could just return it. It's ok not to have this if branch.

File: executor/executor.go [hunk 5]
```diff
@@ -868,6 +879,15 @@ func (e *ExistsExec) Next(ctx context.Context, chk *chunk.Chunk) error {
 	return nil
 }
 
+// Close implements the Executor Close interface.
+func (e *ExistsExec) Close() error {
+	e.childResult = nil
+	if err := errors.Trace(e.baseExecutor.Close()); err != nil {
```
Comment (winoros) [inline executor/executor.go]: ditto.

File: executor/executor.go (hunk 4 above)
Comment (XuHuaiyu) [reply executor/executor.go]: This is what I mean in the previous comments.😂

Comment (lysu) [reply executor/executor.go]: done 😂😂

**Pull request merged**